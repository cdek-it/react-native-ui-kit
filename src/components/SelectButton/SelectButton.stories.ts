import type { Meta, StoryObj } from '@storybook/react'
import { IconArrowDownRight } from '@tabler/icons-react-native'

import { SelectButton } from './SelectButton'

const meta: Meta<typeof SelectButton> = {
  title: 'SelectButton',
  component: SelectButton,
  args: {
    buttons: [
      { label: 'But', Icon: IconArrowDownRight, key: '1' },
      { label: 'ButtonSelect', Icon: IconArrowDownRight, key: '2' },
      { label: 'Button', Icon: IconArrowDownRight, key: '3' },
    ],
    size: 'base',
    initialIndex: 1,
    disabled: false,
  },
  argTypes: {
    size: { control: 'radio', options: ['small', 'base', 'large', 'xlarge'] },
    onPress: { action: 'onPress' },
  },
}

export default meta

type Story = StoryObj<typeof SelectButton>

const SelectButtonStory: Story = {}

export { SelectButtonStory as SelectButton }
