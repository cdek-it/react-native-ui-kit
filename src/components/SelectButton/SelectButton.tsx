import { memo, useCallback, useMemo, useRef, useState } from 'react'
import {
  type AccessibilityProps,
  type LayoutChangeEvent,
  type LayoutRectangle,
  type StyleProp,
  View,
  type ViewProps,
  type ViewStyle,
} from 'react-native'
import Animated, {
  Easing,
  interpolate,
  runOnJS,
  type SharedValue,
  useAnimatedReaction,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

import {
  SelectButtonItem,
  type SelectButtonItemProps,
} from './SelectButtonItem'

const DEFAULT_ANIMATION_DURATION = 200 // ms

export interface SelectButtonProps
  extends AccessibilityProps,
    Pick<ViewProps, 'testID'> {
  /** Массив кнопок. Должен содержать как минимум 2 кнопки. */
  buttons: Array<
    Pick<SelectButtonItemProps, 'label' | 'showIcon' | 'Icon'> & { key: string }
  >

  /** true - если кнопки недоступны для нажатия */
  disabled?: boolean

  /** Индекс выбранной кнопки при первом рендере */
  initialIndex?: number

  /** Вызывается при нажатии на кнопку */
  onPress?: (index: number) => void

  /**
   * Анимированное значение 0...n-1, где n - это количество кнопок.
   * Используется для контроля элемента извне. Если передано, то компонент становится управляемым.
   */
  position?: SharedValue<number>

  /**
   * Выбор размера элемента
   * @default 'base'
   */
  size?: 'small' | 'base' | 'large' | 'xlarge'

  /** Дополнительная стилизация для контейнера компонента */
  style?: StyleProp<ViewStyle>
}

/**
 * Используется для маркировки элементов интерфейса
 * @see https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-4921
 */
export const SelectButton = memo<SelectButtonProps>(
  ({
    buttons,
    disabled,
    initialIndex: initialIndexProp,
    onPress: onPressProp,
    size,
    style,
    testID,
    position: positionProp,
    ...rest
  }) => {
    const styles = useStyles()
    const buttonsLayoutListRef = useRef<LayoutRectangle[]>([])
    const buttonsLayoutList = useSharedValue<LayoutRectangle[]>([])
    const [frameKey, setFrameKey] = useState(Date.now())

    const initialIndex = useMemo(
      () =>
        initialIndexProp && initialIndexProp < buttons.length
          ? initialIndexProp
          : 0,
      [buttons.length, initialIndexProp]
    )

    const onButtonLayout = useCallback(
      (index: number, event: LayoutChangeEvent) => {
        buttonsLayoutListRef.current[index] = event.nativeEvent.layout

        if (buttonsLayoutListRef.current.length === buttons.length) {
          buttonsLayoutList.value = [...buttonsLayoutListRef.current]
        }
      },
      [buttons.length, buttonsLayoutList]
    )

    useAnimatedReaction(
      () => buttonsLayoutList.value,
      (layoutList, previous) => {
        if (layoutList !== previous) {
          runOnJS(setFrameKey)(Date.now())
        }
      }
    )

    const positionInner = useSharedValue(initialIndex)
    const position = useMemo(
      () => positionProp || positionInner,
      [positionInner, positionProp]
    )
    const animationInputRange = useMemo(
      () => buttons.map((_, index) => index),
      [buttons]
    )

    const framePositionStyle = useAnimatedStyle(() => {
      const left = interpolate(
        position.value,
        animationInputRange,
        animationInputRange.map(
          (index) => buttonsLayoutList.value[index]?.x ?? 0
        )
      )
      const width = interpolate(
        position.value,
        animationInputRange,
        animationInputRange.map(
          (index) => buttonsLayoutList.value[index]?.width ?? 0
        )
      )

      return { left, width }
    })

    const isUncontrolledComponent = useMemo(() => !positionProp, [positionProp])
    const onPress = useCallback(
      (index: number) => {
        if (isUncontrolledComponent) {
          positionInner.value = withTiming(index, {
            duration: DEFAULT_ANIMATION_DURATION,
            easing: Easing.linear,
          })
        }

        onPressProp?.(index)
      },
      [isUncontrolledComponent, onPressProp, positionInner]
    )

    return (
      <View
        collapsable={false}
        style={[styles.container, style]}
        testID={testID}
        {...rest}
      >
        {buttons.map(({ label, Icon, key, showIcon }, index) => (
          <SelectButtonItem
            Icon={Icon}
            disabled={disabled}
            index={index}
            key={key}
            label={label}
            position={position}
            showIcon={showIcon}
            size={size}
            testID={`SelectButton_SelectButtonItem_${index}`}
            onLayout={(event) => onButtonLayout(index, event)}
            onPress={() => onPress(index)}
          />
        ))}

        {!disabled && (
          <Animated.View
            key={frameKey}
            style={[styles.frame, framePositionStyle]}
            testID='SelectButton_AnimatedFrame'
          />
        )}
      </View>
    )
  }
)

const useStyles = makeStyles(({ theme }) => ({
  container: {
    flexDirection: 'row',
    padding: theme.Form.SelectButton.selectButtonGroupPadding,
    gap: theme.Form.SelectButton.selectButtonGroupPadding,
    borderRadius: theme.General.borderRadius,
    backgroundColor: theme.Form.SelectButton.selectButtonGroupBg,
  },
  frame: {
    position: 'absolute',
    top: theme.Form.SelectButton.selectButtonGroupPadding,
    bottom: theme.Form.SelectButton.selectButtonGroupPadding,
    left: theme.Form.SelectButton.selectButtonGroupPadding,
    borderRadius: theme.Form.SelectButton.selectButtonBorderRadius,
    backgroundColor: theme.Form.SelectButton.selectButtonActiveBg,
    zIndex: -1,
  },
  textColor: { color: theme.Form.SelectButton.selectButtonTextColor },
  checkedTextColor: {
    color: theme.Form.SelectButton.selectButtonIconActiveColor,
  },
}))
