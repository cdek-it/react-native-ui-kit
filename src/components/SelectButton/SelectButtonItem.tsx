import type { Icon } from '@tabler/icons-react-native'
import { memo, useMemo, useState } from 'react'
import { TouchableOpacity, type ViewProps, type ViewStyle } from 'react-native'
import Animated, {
  interpolateColor,
  runOnJS,
  type SharedValue,
  useAnimatedReaction,
  useAnimatedStyle,
} from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

export interface SelectButtonItemProps
  extends Pick<ViewProps, 'onLayout' | 'testID'> {
  /** Индекс кнопки */
  index: number

  /** Обработчик нажатия на кнопку */
  onPress: () => void

  /**
   * Анимированное значение 0...n-1, где n - это количество кнопок.
   * Кнопка считается выбранной, если значение position равно индексу кнопки.
   */
  position: SharedValue<number>

  /** true - если кнопка недоступна для нажатия */
  disabled?: boolean

  /** Текст на кнопке */
  label?: string

  /**
   * Выбор размера элемента
   * @default 'base'
   */
  size?: 'small' | 'base' | 'large' | 'xlarge'

  /**
   * Показать или скрыть иконку внутри компонента
   * @default true
   */
  showIcon?: boolean

  /** Иконка из набора Tabler */
  Icon?: Icon
}

/**
 * Дочерний элемент компонента SelectButton. Не используется отдельно от SelectButton.
 * @see https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=481-4393
 */
export const SelectButtonItem = memo<SelectButtonItemProps>(
  ({
    index,
    position,
    onPress,
    disabled,
    label,
    onLayout,
    size = 'base',
    showIcon = true,
    Icon,
  }) => {
    const styles = useStyles()

    const iconSize = useMemo(() => {
      switch (size) {
        case 'small':
          return styles.iconSmall

        case 'base':
          return styles.iconBase

        case 'large':
          return styles.iconLarge

        case 'xlarge':
          return styles.iconXLarge
      }
    }, [
      size,
      styles.iconBase,
      styles.iconLarge,
      styles.iconSmall,
      styles.iconXLarge,
    ])

    const labelFontSize = useMemo(() => {
      switch (size) {
        case 'small':
          return styles.labelSmall

        case 'base':
          return styles.labelBase

        case 'large':
          return styles.labelLarge

        case 'xlarge':
          return styles.labelXLarge
      }
    }, [
      size,
      styles.labelBase,
      styles.labelLarge,
      styles.labelSmall,
      styles.labelXLarge,
    ])

    const animatedColorStyle = useAnimatedStyle(() => {
      return {
        color: interpolateColor(
          position.value,
          [index - 1, index, index + 1],
          [
            styles.textColor.color,
            styles.checkedTextColor.color,
            styles.textColor.color,
          ]
        ),
      }
    })

    const [isSelected, setIsSelected] = useState(false)

    useAnimatedReaction(
      () => position.value,
      (value, prevValue) => {
        if (value !== prevValue) {
          runOnJS(setIsSelected)(value === index)
        }
      }
    )

    return (
      <TouchableOpacity
        disabled={disabled}
        style={[
          styles.container,
          styles[size],
          disabled && styles.disabledContainer,
        ]}
        testID='SelectButtonItem_TouchableOpacity'
        onLayout={onLayout}
        onPress={onPress}
      >
        {Icon && showIcon ? (
          <Icon
            height={iconSize.height}
            style={
              [
                styles.textColor,
                isSelected && styles.checkedTextColor,
                disabled && styles.disabledTextColor,
              ] as ViewStyle[]
            }
            testID='SelectButtonItem_Icon'
            width={iconSize.width}
          />
        ) : null}
        <Animated.Text
          numberOfLines={1}
          style={[
            styles.label,
            labelFontSize,
            styles.textColor,
            disabled ? styles.disabledTextColor : animatedColorStyle,
          ]}
          testID='SelectButtonItem_Text'
        >
          {label}
        </Animated.Text>
      </TouchableOpacity>
    )
  }
)

const useStyles = makeStyles(({ theme, typography }) => ({
  container: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    gap: theme.General.inlineSpacing,
    borderWidth: 1,
    borderColor: 'transparent',
  },
  small: { height: theme.Button.Common.buttonHeightSM },
  base: { height: theme.Button.Common.buttonHeight },
  large: { height: theme.Button.Common.buttonHeightLG, gap: 10.5 },
  xlarge: { height: theme.Button.Common.buttonHeightXL, gap: 10.5 },
  disabledContainer: {
    borderRadius: theme.Form.SelectButton.selectButtonBorderRadius,
    borderWidth: 1,
    borderColor: theme.Button.Disabled.disabledButtonBorderColor,
  },
  iconSmall: {
    width: typography.Size['text-base'],
    height: typography.Size['text-base'],
  },
  iconBase: {
    width: typography.Size['text-xl'],
    height: typography.Size['text-xl'],
  },
  iconLarge: {
    width: typography.Size['text-2xl'],
    height: typography.Size['text-2xl'],
  },
  iconXLarge: { width: 28, height: 28 },
  label: { flexShrink: 1, fontWeight: 700 },
  labelSmall: { fontSize: 12.25 },
  labelBase: { fontSize: 14 },
  labelLarge: { fontSize: 17.5 },
  labelXLarge: { fontSize: 21 },
  textColor: { color: theme.Form.SelectButton.selectButtonTextColor },
  checkedTextColor: {
    color: theme.Form.SelectButton.selectButtonIconActiveColor,
  },
  disabledTextColor: { color: theme.Button.Disabled.disabledButtonBorderColor },
}))
