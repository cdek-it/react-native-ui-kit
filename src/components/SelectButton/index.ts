export { SelectButton, type SelectButtonProps } from './SelectButton'
