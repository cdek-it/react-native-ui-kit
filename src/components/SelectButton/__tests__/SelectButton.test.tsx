import { IconArrowDownRight } from '@tabler/icons-react-native'
import {
  fireEvent,
  render,
  userEvent,
  waitFor,
} from '@testing-library/react-native'

import { Pressable } from 'react-native'
import { useSharedValue } from 'react-native-reanimated'

import { SelectButton, type SelectButtonProps } from '../SelectButton'

jest.mock('../SelectButtonItem', () => ({
  SelectButtonItem: 'SelectButtonItem',
}))

type SelectButtonTestProps = Omit<SelectButtonProps, 'buttons'> & {
  readonly controlled?: boolean
}

const buttons = [
  { label: 'ButtonSelect', Icon: IconArrowDownRight, key: 'button1' },
  {
    label: 'ButtonSelect',
    Icon: IconArrowDownRight,
    key: 'button2',
    showIcon: false,
  },
]

const TestComponent = ({ controlled, ...rest }: SelectButtonTestProps) => {
  const position = useSharedValue(0)

  return (
    <>
      <SelectButton
        buttons={buttons}
        position={controlled ? position : undefined}
        {...rest}
      />
      {controlled ? (
        <Pressable
          testID='ChangePosition'
          onPress={() => {
            position.value += 0.5
          }}
        />
      ) : null}
    </>
  )
}

describe('SelectButtonItem', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
    jest.useRealTimers()
  })

  const snapshotCases: Array<[SelectButtonTestProps]> = [
    [{}],
    [{ size: 'small', disabled: true, initialIndex: 1, style: { margin: 10 } }],
    [{ size: 'base', initialIndex: 2 }],
    [{ size: 'large' }],
    [{ size: 'xlarge', initialIndex: 1 }],
  ]

  test.each(snapshotCases)('%p', (props) => {
    const { toJSON, getByTestId } = render(<TestComponent {...props} />)

    buttons.forEach((_, index) => {
      fireEvent(
        getByTestId(`SelectButton_SelectButtonItem_${index}`),
        'layout',
        { nativeEvent: { layout: { width: 100, x: 100 * index } } }
      )
    })

    expect(toJSON()).toMatchSnapshot()
  })

  test('animated frame should be when disabled = true', async () => {
    const { findByTestId } = render(<TestComponent disabled={false} />)

    await expect(
      findByTestId('SelectButton_AnimatedFrame')
    ).resolves.toBeDefined()
  })

  test('animated frame should not be when disabled = false', async () => {
    const { findByTestId } = render(<TestComponent disabled />)

    await expect(
      findByTestId('SelectButton_AnimatedFrame')
    ).rejects.toBeDefined()
  })

  test('handle press on uncontrolled component', async () => {
    const mockedOnPress = jest.fn()
    const { getByTestId } = render(<TestComponent onPress={mockedOnPress} />)
    const selectButtonItem = getByTestId('SelectButton_SelectButtonItem_1')

    buttons.forEach((_, index) => {
      fireEvent(
        getByTestId(`SelectButton_SelectButtonItem_${index}`),
        'layout',
        { nativeEvent: { layout: { width: 100, x: 100 * index } } }
      )
    })

    let animatedFrame = getByTestId('SelectButton_AnimatedFrame')
    await waitFor(() =>
      expect(animatedFrame).toHaveAnimatedStyle({ left: 0, width: 100 })
    )

    selectButtonItem.props.onPress()
    animatedFrame = getByTestId('SelectButton_AnimatedFrame')

    expect(mockedOnPress).toHaveBeenCalledWith(1)

    await waitFor(() =>
      expect(animatedFrame).toHaveAnimatedStyle({ left: 100, width: 100 })
    )
  })

  test('handle press on controlled component', async () => {
    const mockedOnPress = jest.fn()
    const { getByTestId } = render(
      <TestComponent controlled onPress={mockedOnPress} />
    )
    const selectButtonItem = getByTestId('SelectButton_SelectButtonItem_1')

    buttons.forEach((_, index) => {
      fireEvent(
        getByTestId(`SelectButton_SelectButtonItem_${index}`),
        'layout',
        { nativeEvent: { layout: { width: 100, x: 100 * index } } }
      )
    })

    selectButtonItem.props.onPress()

    let animatedFrame = getByTestId('SelectButton_AnimatedFrame')
    await waitFor(() =>
      expect(animatedFrame).toHaveAnimatedStyle({ left: 0, width: 100 })
    )

    const testButton = getByTestId('ChangePosition')
    const user = userEvent.setup()

    await user.press(testButton)
    animatedFrame = getByTestId('SelectButton_AnimatedFrame')

    expect(mockedOnPress).toHaveBeenCalledWith(1)

    await waitFor(() =>
      expect(animatedFrame).toHaveAnimatedStyle({ left: 50, width: 100 })
    )
  })
})
