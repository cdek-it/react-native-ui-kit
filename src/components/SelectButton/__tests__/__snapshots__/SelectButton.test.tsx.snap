// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectButtonItem {"disabled": true, "initialIndex": 1, "size": "small", "style": [Object]} 1`] = `
<View
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 7,
        "flexDirection": "row",
        "gap": 3.5,
        "padding": 3.5,
      },
      {
        "margin": 10,
      },
    ]
  }
>
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    disabled={true}
    index={0}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 1,
      }
    }
    size="small"
    testID="SelectButton_SelectButtonItem_0"
  />
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    disabled={true}
    index={1}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 1,
      }
    }
    showIcon={false}
    size="small"
    testID="SelectButton_SelectButtonItem_1"
  />
</View>
`;

exports[`SelectButtonItem {"initialIndex": 1, "size": "xlarge"} 1`] = `
<View
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 7,
        "flexDirection": "row",
        "gap": 3.5,
        "padding": 3.5,
      },
      undefined,
    ]
  }
>
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={0}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 1,
      }
    }
    size="xlarge"
    testID="SelectButton_SelectButtonItem_0"
  />
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={1}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 1,
      }
    }
    showIcon={false}
    size="xlarge"
    testID="SelectButton_SelectButtonItem_1"
  />
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": 0,
          "width": 0,
        },
      }
    }
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 3.5,
        "bottom": 3.5,
        "left": 0,
        "position": "absolute",
        "top": 3.5,
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="SelectButton_AnimatedFrame"
  />
</View>
`;

exports[`SelectButtonItem {"initialIndex": 2, "size": "base"} 1`] = `
<View
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 7,
        "flexDirection": "row",
        "gap": 3.5,
        "padding": 3.5,
      },
      undefined,
    ]
  }
>
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={0}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 0,
      }
    }
    size="base"
    testID="SelectButton_SelectButtonItem_0"
  />
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={1}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 0,
      }
    }
    showIcon={false}
    size="base"
    testID="SelectButton_SelectButtonItem_1"
  />
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": 0,
          "width": 0,
        },
      }
    }
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 3.5,
        "bottom": 3.5,
        "left": 0,
        "position": "absolute",
        "top": 3.5,
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="SelectButton_AnimatedFrame"
  />
</View>
`;

exports[`SelectButtonItem {"size": "large"} 1`] = `
<View
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 7,
        "flexDirection": "row",
        "gap": 3.5,
        "padding": 3.5,
      },
      undefined,
    ]
  }
>
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={0}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 0,
      }
    }
    size="large"
    testID="SelectButton_SelectButtonItem_0"
  />
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={1}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 0,
      }
    }
    showIcon={false}
    size="large"
    testID="SelectButton_SelectButtonItem_1"
  />
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": 0,
          "width": 0,
        },
      }
    }
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 3.5,
        "bottom": 3.5,
        "left": 0,
        "position": "absolute",
        "top": 3.5,
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="SelectButton_AnimatedFrame"
  />
</View>
`;

exports[`SelectButtonItem {} 1`] = `
<View
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 7,
        "flexDirection": "row",
        "gap": 3.5,
        "padding": 3.5,
      },
      undefined,
    ]
  }
>
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={0}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 0,
      }
    }
    testID="SelectButton_SelectButtonItem_0"
  />
  <SelectButtonItem
    Icon={
      {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    index={1}
    label="ButtonSelect"
    onLayout={[Function]}
    onPress={[Function]}
    position={
      {
        "_isReanimatedSharedValue": true,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 0,
      }
    }
    showIcon={false}
    testID="SelectButton_SelectButtonItem_1"
  />
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": 0,
          "width": 0,
        },
      }
    }
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 3.5,
        "bottom": 3.5,
        "left": 0,
        "position": "absolute",
        "top": 3.5,
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="SelectButton_AnimatedFrame"
  />
</View>
`;
